ENH: Make views produced by np.einsum writeable

insertinterestingnamehere opened the pull request:
`einsum` can be useful for extracting a view of different diagonals of an array, but, currently, the views it returns are not writeable. This changes that.

insertinterestingnamehere committed:
ENH: Make einsum return a writeable view whenever the input is writeable.

insertinterestingnamehere committed:
TST: Test that, when returning a view, einsum returns a writable view
whenever the input array is writable.

charris commented:
Interesting idea, although I'm not sure what the implications would be for writing to the result. @mwiebe Thoughts?

mwiebe commented:
+1 This change looks good to me. I think this is something I just didn't think about when implementing that code.

charris commented:
OK, let's put this in. @insertinterestingnamehere Could you add an entry to `doc/release/1.10.0-notes.rst`.

insertinterestingnamehere commented:
Certainly. I'll do that now.

insertinterestingnamehere committed:
DOC: Add note about writeable views from einsum to 1.10.0 release notes.

charris commented:
Does this only work for single array inputs? An example would be informative.

insertinterestingnamehere commented:
As I understand this, yes, it only works for single array inputs.
Here are some examples of the sort of things this makes possible:

    # working with diagonals
    a = np.random.rand(4, 4, 4)
    # multiply a[0,0], a[1,1], ..., a[4,4] by 10
    np.einsum('iii->i', a) *= 10

    # swapping various dimensions
    a = np.random.rand(3, 3, 3, 3, 3)
    b = np.einsum('abcde->aebcd', a)

Both of the previous operations return views into the original array that, as near as I can tell, are not easily expressed in other ways. The first one requires taking a writeable view of a diagonal twice, and the second would require several clever applications of `np.swapaxes`.

Is there anywhere else you'd like this documented? As near as I can tell, the only thing that is documented right now is that you can return a view at all. IMHO, returning a writeable view seems to be more consistent with the current documentation than returning a read-only one.

insertinterestingnamehere commented:
Hmm, my mistake, the last example can be taken care of fairly easily by `np.rollaxis`. Something like:

    a = np.random.rand(3, 3, 3, 3, 3)
    b = np.einsum('abcde->ebcad', a)

seems to be a little harder though.

charris commented:
Thanks, the examples help a lot. Now that I look, einsum could use some documentation along the lines of a tutorial in `doc/source/reference`. As is, all we have is the documentation in the docstring. We probably need to add something in the docstring along with a `.. versionadded:: 1.10.0` marker, maybe in the Notes section, along with an example with a comment having the numpy version.

insertinterestingnamehere committed:
DOC: Added a note about writeability of views from np.einsum to
the corresponding docstring.
Also added an example showing how to write to the diagonal of an array.

insertinterestingnamehere commented:
@charris I really like the idea of a longer tutorial for this. It's a very useful function that, at first glance, is difficult to understand. I'd be willing to write a tutorial for it at some point, but I probably won't be able to spend much time on it in the immediate future.

insertinterestingnamehere commented:
Also, I just realized the name of the branch is "einsum_readable" not "einsum_writeable". Would you like a new PR for the sake of the git history, or is this good enough?

charris commented:
Don't worry about the branch name.

Yeah, given the complexity of einsum, a tutorial would be nice. You don't have to write it, though )

jaimefrio commented:
Not to take away from the value of this PR, which I am completely +1 on, but the axis rearranging can also be done with `np.transpose`,

```
b = np.einsum('abcde->ebcad', a)
```

is the same as 

```
b = np.transpose(a, (4, 1, 2, 0, 3))
```

insertinterestingnamehere commented:
@jaimefrio Thanks. That is really cool. I wasn't aware that `np.transpose` could be used like that.

insertinterestingnamehere commented:
I think this is done? Is there anything else that I need to take care of?

charris commented:
Yep, looks ready to go. Thanks.
